### DISCUSSION TITLE
[Help wanted] In useFieldArray on "Select" change -> extend with extra data

### DISCUSSION THREAD

**Discussion by ssuvorov** [react-hook-form/react-hook-form#2709] [Q&A] [answered]
Hey guys,
I'd like to extend a field's object on Select change with extra data.

```js
const data = [{
     uuid: 1,
     label: "label1",
     type: "text",
     placeholder,
     salesforce,
     title
  }, {
     uuid: 2,
     label: "label2",
     type: "text2",
     placeholder2,
     salesforce2,
     title2
}]

const defaultValues = {
  fields: [data[0]]
};
```
In `<Select />` I would like to provide a current value - e.g. 1.
**onChange** value to 2, I would like to have all of the fields from `data[1]` object in **fields**.

I’m trying to change and hit submit. Submit values don't contain all fields.
https://codesandbox.io/s/zen-hellman-yfu76?file=/src/App.js

Use Effect doesn't help, it leads to a loop

**Comment by bluebill1049** [comment, marked as answer]
it's not very clear for me here, did you put setValue in the `dep`?

**Reply by ssuvorov** [reply]
```js
<Select
    options={options} // data
    onChange={(item) => {
      console.log("selected", item);
      // item.value here is a UUID from data object
      const newValue = data.find(
        (dataItem) => dataItem.uuid === item.value
      );
      // newValue is a big object with lots of fields
      console.log(newValue);
      // I want to replace here the whole field[index] with newValue (which is data[selectedUUID])
      return setValue(`fields[${index}]`, newValue);
    }}
    onBlur={onBlur}
    value={selectedValue}
    name={name}
/>

**Reply by ssuvorov** [reply]
It doesn't work. Only `uuid` is set to `field[index]`. I tried to use setValue withing `useEffect` -> leads to a loop

**Reply by bluebill1049** [reply]
with the controller you shouldn't use setValue inside, the controller should take care of the setValue with onChange prop. if you want to change other fields, i would recommend to store value in a hidden input and update them. Controller suppose to control single component state.

**Reply by ssuvorov** [reply]
1. I tried to update the hidden fields too. But what's the right way to update them?
OnChange will fire and then?
2. If I won't use a Controller, but register a Select, will it help?

**Reply by bluebill1049** [reply]
useFieldArray doesn't work with custom register, so yes 1 is the correct way.

**Reply by ssuvorov** [reply]
@bluebill1049 but how? :) Will this work?
```js
<Controller
render={({ onChange, ...rest}) => 
  <Select onChange={(value) => {
    setValue(nameOfHiddenField, 1);
    setValue(nameOfHiddenField2, 2);
    onChange(value);
  }} />
}
...
/>
```

**Reply by bluebill1049** [reply]
This is what i meant: https://codesandbox.io/s/epic-violet-yvgr9?file=/src/App.js

**Reply by ssuvorov** [reply]
I'll try it. Thank you a lot for your time. I really appreciate it! Awesome library! Great community 👍

**Reply by ssuvorov** [reply]
It works. Looks a bit weird. Maybe it'd be nice to have a `setValues(object)` method?
I also tried instead of
```js
                              setValue(`${name}[${index}].uuid`, selectedField.uuid);
                              setValue(`${name}[${index}].salesforce`, selectedField.salesforce);
                              setValue(`${name}[${index}].title`, selectedField.title);
                              setValue(`${name}[${index}].placeholder`, selectedField.placeholder);
                              setValue(`${name}[${index}].type`, selectedField.type);
                              setValue(`${name}[${index}].label`, selectedField.label);
```

do
```js
const keyName = `${name}[${index}]`;
reset({
  [keyName]: selectedField
});
```
Doesn't work like that.

**Reply by bluebill1049** [reply]
turn above to a reusable function, and perhaps have a forEach.

**Reply by ssuvorov** [reply]
Yes, that's prettifying. Thanks anyway, I'm closing this issue.

**Reply by adoel90** [reply]
Many- many thanks @bluebill1049 , my problem was resolved. Before find this solution, i was struggle to solved.

**Comment by ssuvorov** [comment]
@bluebill1049 Okay, I worked with it more and figured out a weird behavior

https://codesandbox.io/s/trusting-bogdan-j7w65?file=/src/App.js

The initial array in default values works as expected.
Then I add more items, but they appear neither in watched nor on submit.

**Reply by bluebill1049** [reply]
you should only watch the array field by it's name.
